public/app/features/alerting/unified/rule-list/PaginatedDataSourceLoader.tsx

```diff
@@ -0,0 +1,108 @@
+import { groupBy } from 'lodash';
+import { useEffect, useMemo, useRef } from 'react';
+
+import { Icon, Stack, Text } from '@grafana/ui';
+import { DataSourceRuleGroupIdentifier, ExternalRulesSourceIdentifier, RuleGroup } from 'app/types/unified-alerting';
+
+import { hashRule } from '../utils/rule-id';
+
+import { DataSourceRuleLoader } from './DataSourceRuleLoader';
+import { DataSourceSection, DataSourceSectionProps } from './components/DataSourceSection';
+import { LazyPagination } from './components/LazyPagination';
+import { ListGroup } from './components/ListGroup';
+import { ListSection } from './components/ListSection';
+import { RuleGroupActionsMenu } from './components/RuleGroupActionsMenu';
+import { usePrometheusGroupsGenerator } from './hooks/prometheusGroupsGenerator';
+import { usePaginatedPrometheusGroups } from './hooks/usePaginatedPrometheusGroups';
+
+export const GROUP_PAGE_SIZE = 40;
+
+interface PaginatedDataSourceLoaderProps extends Required<Pick<DataSourceSectionProps, 'application'>> {
+  rulesSourceIdentifier: ExternalRulesSourceIdentifier;
+}
+export function PaginatedDataSourceLoader({ rulesSourceIdentifier, application }: PaginatedDataSourceLoaderProps) {
+  const { uid, name } = rulesSourceIdentifier;
+  const prometheusGroupsGenerator = usePrometheusGroupsGenerator();
+
+  const groupsGenerator = useRef(prometheusGroupsGenerator(rulesSourceIdentifier, GROUP_PAGE_SIZE));
+
+  useEffect(() => {
+    const currentGenerator = groupsGenerator.current;
+    return () => {
+      currentGenerator.return();
+    };
+  }, [groupsGenerator]);
+
+  const {
+    page: groupsPage,
+    nextPage,
+    previousPage,
+    canMoveForward,
+    canMoveBackward,
+    isLoading,
+  } = usePaginatedPrometheusGroups(groupsGenerator.current, GROUP_PAGE_SIZE);
+
+  const groupsByNamespace = useMemo(() => groupBy(groupsPage, 'file'), [groupsPage]);
+
+  return (
+    <DataSourceSection name={name} application={application} uid={uid} isLoading={isLoading}>
+      <Stack direction="column" gap={1}>
+        {Object.entries(groupsByNamespace).map(([namespace, groups]) => (
+          <ListSection
+            key={namespace}
+            title={
+              <Stack direction="row" gap={1} alignItems="center">
+                <Icon name="folder" />{' '}
+                <Text variant="body" element="h3">
+                  {namespace}
+                </Text>
+              </Stack>
+            }
+          >
+            {groups.map((group) => (
+              <RuleGroupListItem
+                key={`${String(rulesSourceIdentifier.uid)}-${namespace}-${group.name}`}
```

Comment:
Do we want to/should we reuse the `getRuleKey` method from the FilterView? I know you've left code comments about being careful when we compute that though

Reply:
I think we should try to colocate whenever we can, unless the key computation for rule groups is something we need in other places.

Reply:
No, this is a key for a group, which is fairly simple - 3 concatenated strings. 
`getRuleKey` computes a key for a rule